pkg/apis/flowcontrol/types.go

```diff
@@ -0,0 +1,298 @@
+// intermediate conditions, e.g. ConditionDegraded.
+const (
+	FlowSchemaConditionTrue    FlowSchemaConditionStatus = "True"
+	FlowSchemaConditionFalse   FlowSchemaConditionStatus = "False"
+	FlowSchemaConditionUnknown FlowSchemaConditionStatus = "Unknown"
+
+	PriorityLevelConfigurationConditionTrue    PriorityLevelConfigurationConditionStatus = "True"
+	PriorityLevelConfigurationConditionFalse   PriorityLevelConfigurationConditionStatus = "False"
+	PriorityLevelConfigurationConditionUnknown PriorityLevelConfigurationConditionStatus = "Unknown"
+)
+
+const (
+	FlowDistinguisherMethodByUserName  FlowDistinguisherMethod = "ByUserName"
+	FlowDistinguisherMethodByUserGoup  FlowDistinguisherMethod = "ByUserGroup"
+	FlowDistinguisherMethodByNamespace FlowDistinguisherMethod = "ByNamespace"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// FlowSchema defines the schema of a group of flows. Note that a flow makes up of a set of inbound API requests with
+// similar attributes and is identified by a pair of strings: the name of the FlowSchema and a "flow distinguisher".
+type FlowSchema struct {
+	metav1.TypeMeta
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+	// Specification of the desired behavior of a flow-schema.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Spec FlowSchemaSpec
+	// Current status of a flow-schema.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Status FlowSchemaStatus
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// FlowSchemaList is a list of FlowSchema objects.
+type FlowSchemaList struct {
+	metav1.TypeMeta
+	// Standard list metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta
+
+	// Items is a list of flow-schemas.
+	Items []FlowSchema
+}
+
+// FlowSchemaSpec describes how the flow-schema's specification looks like.
+type FlowSchemaSpec struct {
+	// PriorityLevelConfigurationRef can reference a PriorityLevelConfiguration in the cluster. If the reference cannot
+	// be resolved, the flow-schema will ignored and marked as invalid in its status.
+	PriorityLevelConfigurationRef PriorityLevelConfigurationRef
+	// MatchingPrecedence is used to choose among the FlowSchemas that match a given request.  The chosen
+	// FlowSchema is among those with the numerically lowest (which we take to be logically highest)
+	// MatchingPrecedence.  Each MatchingPrecedence value must be non-negative.
+	MatchingPrecedence int32
+	// DistinguisherMethod defines how to classify requests into certain flows. If not empty, the flow's identifier is computed
+	// according to specification of the distinguisher.
```

Comment:
This begs the question, what if this field is unspecified (i.e., at its zero value)?

Reply:
what about a "None" method as zero/unspecified value? at the "None" case,  all the requests matching the flow-schema will be pushed into one single queue and the shuffle-sharding won't be working?

Reply:
Yes, that sounds good.  That will require an adjustment to `FlowDistinguisherMethodNone` above, it current is _not_ defined to be the empty string.

Reply:
added "none" method